controller/resource.go

```diff
@@ -457,15 +541,51 @@ func (t *TaskState) Delete() error {
 	return nil
 }
 
+// DeregisterService removes the service from the Consul catalog
+func (t *TaskState) DeregisterService(consulClient *api.Client) error {
+	opts := &api.WriteOptions{Partition: t.Partition, Namespace: t.Service.Namespace}
+	deregInput := &api.CatalogDeregistration{
+		Node:      t.ClusterARN,
+		ServiceID: t.Service.ID,
+		Partition: t.Partition,
+		Namespace: t.Service.Namespace,
+	}
+
+	_, err := consulClient.Catalog().Deregister(deregInput, opts)
+	if err != nil {
+		return fmt.Errorf("deregistering service with ID %s: %w", t.Service.ID, err)
+	}
+	return nil
+}
+
 // Namespace returns the namespace that the service belongs to.
 // It returns the empty string if namespaces are not enabled.
 func (t *TaskState) Namespace() string {
-	if t.ECSTaskFound {
+	if t.IsPresent() {
 		return t.NS
 	}
 	return ""
 }
 
+// ID returns the taskID of the resource
+func (t *TaskState) ID() TaskID {
+	return t.TaskID
+}
+
+// IsACLNotFoundError returns true if the ACL is not found.
+func IsACLNotFoundError(err error) bool {
+	return err != nil && strings.Contains(err.Error(), "Unexpected response code: 403 (ACL not found)")
```

Comment:
minor: We may want to change this check to just `strings.Contains(err.Error(), "ACL not found")`, [similar to what Consul has defined](https://github.com/hashicorp/consul/blob/efe981637b44d9e154474de2ba5ca6e6ecb4ca8d/acl/errors.go#L43-L47).

There were some ACL response code and error message changes in Consul 1.15, so I'm suspicious of comparing with a longer string here.

Non-blocking because you haven't changed this, but something to keep in mind.